### DISCUSSION TITLE
vsgImGui: How to support InputMethod?

### DISCUSSION THREAD

**Discussion by duck-counter** [vsg-dev/VulkanSceneGraph#1319] [Q&A] [answered]
Hello, everyone!

The examples of vsgImGui don’t demonstrate how to support input methods. I’d like to know how to achieve this.
I’m trying to input Chinese characters in this text box using a pinyin input method.
There are two issues to resolve:
1. Only the pinyin is retained in the text box, but I need it to keep the Chinese characters.
2. The input method window is currently positioned in the top-left corner of the screen, but it needs to follow the cursor position.
![image](https://github.com/user-attachments/assets/6fcdb5db-a23b-47c5-a178-b58f346ba5ea)

The effect I hope to achieve is something like this:
![image](https://github.com/user-attachments/assets/11fc9b47-22fd-4db3-b2d3-c3467458cecb)

My current thought is that VSG needs to pass more messages to ImGui, but I don’t quite understand VSG’s messaging mechanism yet. It seems that VSG has some encapsulation in this area. If you could provide an overview of VSG’s message handling design, it would be a great help to me.

Thank you in advance!

**Comment by robertosfield** [comment, marked as answer]
The VSG doesn't have a "messaging mechanism".

Do you mean platform agnostic event encapsulation?  It just provides a
means of encapsulating platform specific Windows/Keyboard/Mouse events.
Events are polled each frame and put into a event list than applications
can capture with event handler as is done by [vsgImGui::SendEventsToImGui(
https://github.com/vsg-dev/vsgImGui/blob/master/include/vsgImGui/SendEventsToImGui.h
)

As for InputMethod, again having to guess what you mean.  In this in
context of ImGui?  If so what API are you referring to? Could you provide a
code example of what you mean as I'm not an expert in ImGui and don't have
direct experience with working with Chinese characters sets.

**Reply by duck-counter** [reply]
> The VSG doesn't have a "messaging mechanism". Do you mean platform agnostic event encapsulation? It just provides a means of encapsulating platform specific Windows/Keyboard/Mouse events. Events are polled each frame and put into a event list than applications can capture with event handler as is done by [vsgImGui::SendEventsToImGui( https://github.com/vsg-dev/vsgImGui/blob/master/include/vsgImGui/SendEventsToImGui.h ) As for InputMethod, again having to guess what you mean. In this in context of ImGui? If so what API are you referring to? Could you provide a code example of what you mean as I'm not an expert in ImGui and don't have direct experience with working with Chinese characters sets.
> […](#)
> Message ID: ***@***.***>

Got it, thanks a lot.
Based on your suggestion, I did some further searching, and this is another clue: 
https://github.com/ocornut/imgui/blob/master/docs/FAQ.md
![image](https://github.com/user-attachments/assets/ea5db60a-4d7e-4d41-8c81-33cf134ab040)

So, what I might need to do is:
1. Handle WM_UNICHAR to get the characters input by the user;
2. Pass these characters to ImGui's io.AddInputCharacter();
3. Set the IME position, although I'm not entirely sure how yet, but I should be able to refer to ImGui's Win32 example program.

I will try it over the weekend.

**Comment by robertosfield** [comment]
The relevant code in vsgImGui is:

~~~ cpp
    if (uint16_t c = keyPress.keyModified; c > 0 && c < 255)
    {
        io.AddInputCharacter(c);
    }
~~~

The c>0 && c<255 will be due to ImGui not directly handling non ascii characters, the VSG itself will be handling them.

So my guess is that we just need to add some code to convert the non ascii characters to UTF-8.
